File: DataStructures/HashTables/HashTable.rb
=begin
 * Hash Table Implemented from scratch via Ruby
=end

class HashTable
  attr_accessor :size, :store

  def initialize(size = 10)
    @size = size
    @store = [] #Array.new(size) { nil }
  end

  def hash(key)
    hash = 0

    key.each_byte do |x|
      hash = (hash << 5) + hash + x
      hash = hash.floor.abs
    end

    return hash % @size
  end

  def insert(key, value)
    index = hash(key)

    if @store[index] === nil
      @store[index] = [[key, value]]
    else
      inserted = False

      @store[index].each do |x|
        if x[0] === key
          x[1] = value
        end
      end

      if inserted === false
        @store[index].push([key, value])
      end
    end
  end

  def retrieve(key)
    index = hash(key)

    if @store[index] === nil
      return nil
    else

      @store[index].each do |x|
        if x[0] === key
          return x[1]
        end
      end
    end
  end
end

=begin
  * HashTable using the Ruby Standard Library
=end

table = { jenny: 'Ugly' }
table[:bobby] = 'Fat'
table[:frank] = 'Smelly'

# table[:bobby] - access a value with this key
# table.clear - remove all key/value pairs
# table.default - set a default value for a key that doesn't exist
# table.delete('frank') - delete key/value pair and return vbalue
# table.each_/key/pair/value - iterate over different data in hash

File: DataStructures/HashTables/HashTable.js
/**
 * Hash Table Implemented from scratch via JavaScript
 * Helpful: http://www.willvillanueva.com/javascript-hash-tables/
 */


function HashTable(limit = 10) {
 this.store = [];
 this.limit = limit;
}

/* Prototype function that converts a key (string) in a numerical value
 * based on the contents of that key */
HashTable.prototype.hash = function hash(key) {
 var hash = 0;
 var len = key.length;

 for (var x = 0; x < len; x++) {
   hash = (hash << 5) + hash + key.charCodeAt(x);
   hash = hash | 0; // bitwise conversion to integer
   hash = Math.abs(hash);
 }

 return hash % this.limit;
};

HashTable.prototype.insert = function insert(key, value) {
  var index = this.hash(key);

  if (!this.store[index]) this.store[index] = [[key, value]];
  else {
    var inserted = false;
    var len = this.store[index].length;

    for (var x = 0; x < len; x++) {
      if (this.store[index][x][0] === key) {
        this.store[index][x][1] = value;
        inserted = true;
      }
    }

    if (!inserted) this.store[index].push([key, value]);
  }
};

HashTable.prototype.retrieve = function retrieve(key) {
  var index = this.hash(key);

  if (!this.store[index]) return undefined;
  else {
    var len = this.store[index].length;

    for (var x = 0; x < len; x++) {
      if (this.store[index][x][0] === key) {
        return this.store[index][x][1];
      }
    }
  }
};

/**
 * A HashTable using JavaScript's Standard Library object
 */
var table = {};
table.apple = 'red';
table['blueberry'] = 'blue';

// table.freeze() - no code can alter properties
// table.keys() - returns array caontaining names of object's properties (keys)
// table.hasOwnProperty(key) - returns bool indicating existance of property (key)
// table.toString() - returns a string of the table
